rfcs/2020-10-29-3736-remap-chain-parsing.md

```diff
@@ -0,0 +1,215 @@
+condition will be valid for both a single (`=`) and double (`==`) equals in the
+condition, yet both variations do very different things:
+
+```
+if .foo == true {
+
+}
+```
+
+```
+if .foo = true {
+
+}
+```
+
+This has haunted C programmers for decades.
+
+Also, the assignment still occurs even if the condition doesn't succeed. Side
+effects such as this can be unexpected for the script writer and could cause
+unexpected bugs to creep in.
+
+
+## Alternatives
+
+### Enforce boolean conditions
+Allowing for non-boolean conditions in the `if` condition is not strictly
+necessary. A similar result could be achieved by running the regular expression
+twice, one with `match` which returns a boolean if it matches, and then again
+with `matches` which would extract the matches out of the text.
+
+```coffeescript
+if match(.message, /^Started (?P<method>[^\s]*) for (?P<remote_addr>[^\s]*)/) {
+  $match = matches(.message, /^Started (?P<method>[^\s]*) for (?P<remote_addr>[^\s]*)/)
+  .method = $match.method
+  .remote_addr = $match.remote_addr
+  .source = "nginx"
+}
+```
+
+However, this has performance implications as the regular expression has to be
+run twice.
+
+### Use `switch`
+There are some alternative syntaxes that may be worth considering. One
+possibility that is common in many languages is a `switch` statement.
+
+```coffeescript
+switch {
+ ($match = matches(.message, /^Started (?P<method>[^\s]*) for (?P<remote_addr>[^\s]*)/)): {
+    .method = $match.method
+    .remote_addr = $match.remote_addr
+    .source = "nginx"
+  }
+ ($match = matches(.message, /^(?P<remote_addr>[^\s]*).*"(?P<method>[^\s]*).*"$/): {
+    .method = $match.method
+    .remote_addr = $match.remote_addr
+    .source = "haproxy"
+  }
+  default: {
+    .source = "other"
+  }
+}
+```
+
+The disadvantage of this approach is it is an extra syntactical element that
+needs to be added to the language that would require maintenance and make the
+language harder to learn. The advantage, however, is that we can then
+potentially avoid having to enable `if` statements to work with non-boolean
+values. However, I'm not convinced this would be a good idea as having
+conditions working differently depending on their context could be confusing for
+the user.
+
+
+## Outstanding Questions
+
+- One thing that the `swimlanes` can do that this RFC hasn't catered for is that
+`swimlanes` can send the event to different sinks depending on a condition. With
+Remap there is only one possible component, another transform or a sink, that
+can accept the output of the mapping. Is this something that should be catered
+for by the Remap language and this RFC?
+
+
+## Plan Of Attack
+
+Incremental steps that execute this change. Generally this is in the form of:
+
+- [ ] Add `else if` to the remap language parser and add multiple branches to
+      the `IfStatement` struct to handle multiple conditions.
+- [ ] Change the `IfStatement` functionality to treat non-boolean values as pass
```

Comment:
If we do this, this seems like it would also have implications for the `reduce` transform in that we should match the behavior there for how the end result is interpreted per the discussion starting with https://github.com/timberio/vector/pull/4743#pullrequestreview-517050895

I'm starting to feel like we need to make a decision between handling types strictly or loosely, generally, to have a consistent approach. I feel like strict typing would have all of the references to incoming data tagged as "any" which users could then refine down through type coercion functions.

Reply:
I lean towards strictly. A principle of this language is safety.

Reply:
Good call. I will update the RFC to include this and incorporate @JeanMertz  suggestion of grouping multiple statements.

Reply:
I feel like we'll still need to do some work if we want that check to happen at compile-time though.

For example, if I do:

```
if .foo {
  $bar = "baz"
}
```

The type of `.foo` is dependent on the input. Would we require `if .foo == true { }`?

Reply:
Type checking happens at runtime (since we can't know event field types at compile-time), but that doesn't mean we are treating types loosely.

Your example can result in an error for one event, and a successful execution for another, depending on the value their `.foo` field contains. This is also why we have coercion functions such as `to_bool`.

If you want to make sure the condition you mentioned doesn't fail, you'll need to make sure the value resolves to a boolean, either through `to_bool` or any other function that returns a boolean, or through any binary operation such as `.foo == true`.

This is also why we're considering providing safe functions such as `parse_json!` vs `parse_json`, such that `if parse_json` won't result in an error, but skip the if-condition if JSON parsing failed. This won't work for built-in operators though, but we can add a wrapper function such as `safe` which takes any expression, and returns a `Err` into a `Ok(None)`, which is what the safe functions also do.

To go back to the `if .foo` example, if you'd were to write it as `if to_bool(.foo)`, and it would receive a value it can't coerce to a bool (e.g. a map), it would return `Ok(None)`, which would function similar to `false` in the if-statement. If you used `if to_bool!(.foo)`, the expression would only succeed if the coercion succeeded, otherwise the script would terminate.

It should be noted that all coercion functions take a secondary "default" argument, so all of this can be avoided by doing `if to_bool(.foo, default = false)`

Reply:
:thinking: 

> If you want to make sure the condition you mentioned doesn't fail, you'll need to make sure the value resolves to a boolean, either through to_bool or any other function that returns a boolean, or through any binary operation such as .foo == true.

By fail do you mean that it resolves to `false` and executes the `else` blocks (if there is one?) or do you mean that it throws an error?

I guess my thinking is that we could introduce even more compile-time safety (perhaps as an optional "strict mode" if we don't want to make it the default), by failing to load a `remap` script where the type used in a conditional is ambiguous (like simply `if .foo { }`); requiring users to wrap it in a `to_bool`. This will prompt them to think about what should happen if `.foo` is, for some reason, not a boolean value and make it clearer that it won't automatically coerce values (which we don't). They could choose to use `to_bool!()` if they know it will always be a boolean.

Reply:
This same compile-time type safety could be extended to situations like `.foo > 5` where it would fail to load unless you ensure that `.foo` to a numeric value through type coercion.

Reply:
I like that idea. It would force the user to think through the assumptions that they are making about the data they are working with.

Reply:
This would be possible (but is not part of) if we implement RFC #4839.

What we'd do in that case is, whenever the `Comparison` expression is created, it checks the `resolves_to` value of both the left-hand side and right-hand side expressions, and determine if both return either an integer or float.